src/Workspaces/CSharpTest/EmbeddedLanguages/RegularExpressions/CSharpRegexParserTests.cs

```diff
@@ -0,0 +1,291 @@
+        }
+
+        private void TryParseSubTrees(string stringText, RegexOptions options)
+        {
+            // Trim the input from the right and make sure tree invariants hold
+            var current = stringText;
+            while (current != "@\"\"" && current != "\"\"")
+            {
+                current = current.Substring(0, current.Length - 2) + "\"";
+                TryParseTree(current, options, conversionFailureOk: true);
+            }
+
+            // Trim the input from the left and make sure tree invariants hold
+            current = stringText;
+            while (current != "@\"\"" && current != "\"\"")
+            {
+                if (current[0] == '@')
+                {
+                    current = "@\"" + current.Substring(3);
+                }
+                else
+                {
+                    current = "\"" + current.Substring(2);
+                }
+
+                TryParseTree(current, options, conversionFailureOk: true);
+            }
+            
+            for (int start = stringText[0] == '@' ? 2 : 1; start < stringText.Length - 1; start++)
+            {
+                TryParseTree(
+                    stringText.Substring(0, start) +
+                    stringText.Substring(start + 1, stringText.Length - (start + 1)),
+                    options, conversionFailureOk: true);
+            }
+        }
+
+        private (SyntaxToken, RegexTree, ImmutableArray<VirtualChar>) JustParseTree(
+            string stringText, RegexOptions options, bool conversionFailureOk)
+        {
+            var token = GetStringToken(stringText);
+            var allChars = _service.TryConvertToVirtualChars(token);
+            if (allChars.IsDefault)
+            {
+                Assert.True(conversionFailureOk, "Failed to convert text to token.");
+                return (token, null, allChars);
+            }
+
+            var tree = RegexParser.TryParse(allChars, options);
+            return (token, tree, allChars);
+        }
+
+        private RegexTree TryParseTree(string stringText, RegexOptions options, bool conversionFailureOk)
+        {
+            var (token, tree, allChars) = JustParseTree(stringText, options, conversionFailureOk);
+            if (tree == null)
+            {
+                Assert.True(allChars.IsDefault);
+                return null;
+            }
+
+            CheckInvariants(tree, allChars);
+
+            try
+            {
+                var regex = new Regex(token.ValueText, options);
+                Assert.Empty(tree.Diagnostics);
+
+                Assert.True(regex.GetGroupNumbers().OrderBy(v => v).SequenceEqual(
+                    tree.CaptureNumbersToSpan.Keys.OrderBy(v => v)));
+
+                Assert.True(regex.GetGroupNames().Where(v => !int.TryParse(v, out _)).OrderBy(v => v).SequenceEqual(
+                    tree.CaptureNamesToSpan.Keys.OrderBy(v => v)));
+            }
+            catch (IndexOutOfRangeException)
```

Comment:
Try/catching seems too loose. Can tests that trigger those cases pass a parameter to signal their expectation?

Reply:
sure. i can do that.

Reply:
Ok.  I did this.  Tests now need to explicitly state if they expect hte regex constructor to throw.  Also, we only wrap creating the .net regex in a try/catch.  That way we're only checking for exceptions in it, and we're not accidentally catching exceptions in the roslyn regex system.